midend/parserUnroll.cpp

```diff
@@ -666,6 +667,11 @@ class ParserSymbolicInterpreter {
                         IR::IndexedVector<IR::StatOrDecl>(),
                         new IR::PathExpression(new IR::Type_State(),
                             new IR::Path(outOfBoundsStateName, false)));
+                } else {
```

Comment:
why is this "else" needed?
I could argue that the program for which this "bug" happens is certainly incorrect, because it will always produce a runtime error in the start state. Are there any "correct" programs where this bug is present?

Reply:
I think that this bug appear if program is incorrect (like issue314.p4), but even in that case parserUnroll can't eliminate states from a parser.

Reply:
This is not really an explanation. My question is "why don't you do the addition in the "if" branch as well.

Reply:
What kind of addition you mean? evaluateState returns nullptr in the first argument if some error appeared during execution. `If body` corresponds to the generation of a `OutOfBound` states and `else body` corresponds to all other cases. `notAdded` flag is used to add old state once only.